### PR TITLE
Bson plugin

### PR DESCRIPTION
**Pull request by TorgeirUstad**
Add BSON plugin for DLite

DLite seems to currently mess up binary strings, by replacing '\\' with '\\\\', which messes up control sequences (e.g. '\n' (hex '0a') becomes '\\\\n', which is interpreted as '\\\\' and 'n' (i.e hex '5c6e')). The plugin is therefore written as a workaround. If this issue can be resolved, the plugin Python code can probably be simplified.

The plugin has only been tested for BSON files generated by itself (by pymongo), not for BSON files generated by other sources.

EDIT: The problem mentioned above has been fixed, so the plugin has been simplified.

### CODE REVIEW

File: storages/python/python-storage-plugins/bson.py [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+        If `writable` is not set, it is assumed to be True.
+        
+        The BSON data is translated to JSON.
+        """
+        self.options = Options(options, defaults='mode=append')
+        self.mode = dict(r='rb', w='wb', append='rb+')[self.options.mode]
+        self.writable = False if 'rb' in self.mode else True
+        self.uri = uri
+        self.d = {}
+        if self.mode in ('rb', 'rb+') and os.path.exists(uri):
```
Comment (jesper-friis) [inline storages/python/python-storage-plugins/bson.py:47]: This will not fail if `uri` doesn't exists - but we want it to fail in this case since we are opening in read mode

File: storages/python/python-storage-plugins/bson.py [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+            with open(uri, self.mode) as f:
+                bson_data = f.read()
+                if pybson.is_valid(bson_data):
```
Comment (jesper-friis) [inline storages/python/python-storage-plugins/bson.py:50]: We can end the with-block here, since are done reading the data

File: storages/python/python-storage-plugins/bson.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+                    self.d = pybson.decode(bson_data)
+                    if self.d == {}:
```
Comment (jesper-friis) [inline storages/python/python-storage-plugins/bson.py:52]:
```suggestion
                    if not self.d:
```

File: storages/python/python-storage-plugins/bson.py [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+                    d = d + "'" + s[(n + 2):]
+                entries[uuid] = d
+            with open(self.uri, mode) as f:
+                f.write(pybson.encode(entries))
+
+    def load(self, uuid):
+        """Load `uuid` from current storage and return it
+        as a new instance.
+        """
+        for key, value in self.d.items():
```
Comment (jesper-friis) [inline storages/python/python-storage-plugins/bson.py:132]: No need to loop through `self.d`. You can get the value with `value = self.d[uuid]`

File: storages/python/python-storage-plugins/bson.py [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+            if key == uuid:
+                if type(value) == str:
+                    value = eval(value) # Translate to dict
```
Comment (jesper-friis) [inline storages/python/python-storage-plugins/bson.py:135]: This is a security issue. Are we 100% sure that the `value` cannot be a malicious designed string?
Eval can execute anything. Safer to import `ast` and write `value = ast.literal_eval(value)`.

File: storages/python/python-storage-plugins/bson.py [hunk 6]
```diff
@@ -0,0 +1,170 @@
...
+                # Create instance and return it
```
Comment (jesper-friis) [inline storages/python/python-storage-plugins/bson.py:136]: Isn't it possible to just return `instance_from_dict(value)` at this point?

File: storages/python/python-storage-plugins/bson.py [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+                        raise EOFError("Failed to read BSON data from " + uri)
+                else:
+                    raise EOFError("Invalid BSON data in source " + uri)
+
+    def close(self):
+        """Close this storage and write the data to file.
+
+        Assumes the data to store is in JSON format.
+        """
+        if self.writable:
```
Comment (jesper-friis) [inline storages/python/python-storage-plugins/bson.py:62]: I am probably an ignorant, but I don't understand why we need these conversions.

I testet the following which works for me (prints True):
```python
import bson

import dlite
from dlite.utils import instance_from_dict


inst = dlite.get_instance(dlite.ENTITY_SCHEMA)

data = bson.encode(inst.asdict())
with open('raw.bson', 'wb') as f:
    f.write(data)

with open('raw.bson', 'rb') as f:
    newdata = f.read()
d = bson.decode(newdata)
inst2 = instance_from_dict(d)

print(inst == inst2)
```

File: storages/python/tests/test_bson_storage.c [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+{
+  DLiteStorage* s = NULL;
+
+  // Load JSON metadata
+  char* url = "json://" STRINGIFY(DLITE_ROOT) "/src/tests/test-entity.json?mode=r";
+  DLiteInstance* json_meta = (DLiteInstance*)dlite_meta_load_url(url);
+  char* json_str1 = dlite_json_aprint(json_meta, 0, 1);
+  while (dlite_instance_decref(json_meta)); // Remove all references to json_meta
+
+  // Load BSON metadata
```
Comment (jesper-friis) [inline storages/python/tests/test_bson_storage.c:52]:
```suggestion
  // Load BSON data
```

File: storages/python/tests/test_bson_storage.c [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+  bson_str1 = dlite_json_aprint(bson_data, 0, 1);
+  while (dlite_instance_decref(bson_data)); // Remove all references to bson_data
+  bson_data = dlite_instance_load(s, "e076a856-e36e-5335-967e-2f2fd153c17d");
+  mu_check(bson_data);
+  char* bson_str2 = dlite_json_aprint(bson_data, 0, 1);
+  while (dlite_instance_decref(bson_data)); // Remove all references to bson_data
+
+  // Compare JSON and BSON data
+  mu_assert_int_eq(0, strcmp(json_str1, bson_str1));
+  mu_assert_int_eq(0, strncmp(json_str2, bson_str2, 47));
```
Comment (jesper-friis) [inline storages/python/tests/test_bson_storage.c:86]: why do you use strncmp and not just strcmp?

File: storages/python/tests/test_bson_storage.c [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+#include "dlite.h"
+#include "dlite-macros.h"
+#include "dlite-storage-plugins.h"
+
+MU_TEST(test_save)
+{
+  DLiteStorage* s = NULL;
+
+  // Load JSON metadata
+  char* url = "json://" STRINGIFY(DLITE_ROOT) "/src/tests/test-entity.json?mode=r";
```
Comment (jesper-friis) [inline storages/python/tests/test_bson_storage.c:12]: In general we would like to keep lines shorter than 80 characters - makes them easy to read in a 80 character wide termical...

File: storages/python/python-storage-plugins/bson.py [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+        self.writable = False if 'rb' in self.mode else True
+        self.uri = uri
+        self.d = {}
+        if self.mode in ('rb', 'rb+') and os.path.exists(uri):
+            with open(uri, self.mode) as f:
+                bson_data = f.read()
+                if pybson.is_valid(bson_data):
+                    self.d = pybson.decode(bson_data)
+                    if self.d == {}:
+                        raise EOFError("Failed to read BSON data from " + uri)
```
Comment (jesper-friis) [inline storages/python/python-storage-plugins/bson.py:53]: It is good practice to keep lines below 80 characters

File: storages/python/tests/test_bson_storage.c (hunk 9 above)
Comment (jesper-friis) [reply storages/python/tests/test_bson_storage.c:86]: Ahh, ser nå hvorfor. bson-parseren setter ikke `uri`. Det er vel egentlig noe vi ønsker at den skal gjøre.

**Review by jesper-friis — COMMENTED** [review]
Hi, I have added some comments alongside the code

File: storages/python/tests/test_bson_storage.c (hunk 10 above)
Comment (TorgeirUstad) [reply storages/python/tests/test_bson_storage.c:12]: I know, but I saw that it wasn't practiced throughout the DLite source code, so I didn't pay attention to it. But I'll fix it.

File: storages/python/python-storage-plugins/bson.py (hunk 5 above)
Comment (TorgeirUstad) [reply storages/python/python-storage-plugins/bson.py:135]: I know. I looked at ast.literal_eval(), but decided against using it due to the warning:
https://docs.python.org/3/library/ast.html?highlight=ast%20literal_eval#ast.literal_eval

I'm not sure if the issue stated in the warning would be an issue or not in practical use of the plugin.

File: storages/python/python-storage-plugins/bson.py (hunk 11 above)
Comment (TorgeirUstad) [reply storages/python/python-storage-plugins/bson.py:53]: I'll fix it (see reply above).

File: storages/python/python-storage-plugins/bson.py (hunk 7 above)
Comment (TorgeirUstad) [reply storages/python/python-storage-plugins/bson.py:62]: Is this metadata? The problem I encountered was with actual data (the 'myblob' properties in the file 'src/tests/test-data.json'). The DLite entity stores e.g. the 'myblob' value '0a0010' (for uuid 'e076a856-e36e-5335-967e-2f2fd153c17d') as "bytearray(b'\n\x00\x10')", and bson.encode() didn't accept that input, so it had to be translated. But translating it directly back to '0a0010' didn't work**, as DLite somehow caused the backslashes to be doubled, so that the translated hexadecimal value became '5c6e5c7830305c783130' (corresponding to b'\\\n\\\x00\\\x10').

** It worked when run through the Python interpreter, but not when run through DLite. This is why I wrote "The plugin is therefore written as a workaround. If this issue can be resolved, the plugin Python code can probably be simplified."

File: storages/python/python-storage-plugins/bson.py (hunk 6 above)
Comment (TorgeirUstad) [reply storages/python/python-storage-plugins/bson.py:136]: For metadata, yes, For actual data, it turned out that I had to handle the hexadecimal input explicitly.

File: storages/python/python-storage-plugins/bson.py (hunk 1 above)
Comment (TorgeirUstad) [reply storages/python/python-storage-plugins/bson.py:47]: Ok. Fail, as in raising an exception?

File: storages/python/python-storage-plugins/bson.py (hunk 5 above)
Comment (jesper-friis) [reply storages/python/python-storage-plugins/bson.py:135]: You are right, `ast.literal_eval()` does not protect against denial-of-service attacks, but it is much better than `eval()`. I definitely think you should use ast.literal_eval().

File: storages/python/python-storage-plugins/bson.py (hunk 1 above)
Comment (jesper-friis) [reply storages/python/python-storage-plugins/bson.py:47]: Yes, raising an exception with a good message would be the preferred behaviour

File: storages/python/python-storage-plugins/bson.py (hunk 6 above)
Comment (jesper-friis) [reply storages/python/python-storage-plugins/bson.py:136]: Ahh, I see. I think the right solution would be to fix `instance_from_dict()` rather than you having to handle all the edge cases. Please add an issue.

File: storages/python/tests/test_bson_storage.c (hunk 10 above)
Comment (jesper-friis) [reply storages/python/tests/test_bson_storage.c:12]: Probably because there have been several authors. I think we should standardise on 80 character line length.

**Review by jesper-friis — APPROVED** [review]
